### ISSUE TITLE
tearDownAll callback runs even if no tests run

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/test#1422]
The `tearDownAll` documentation states:

> If this is called within a test group, `callback` will run after all tests in that group. It will be run before any tearDownAll callbacks in parent groups or at the top level. It won't be run if none of the tests in the group are run.

Consider the following code:

```dart
import 'package:test/test.dart';

void main() {
  group('foo', () {
    setUpAll(() {
      print('setUpAll');
    });

    tearDownAll(() {
      print('tearDownAll');
    });
  });
}
```
There are no tests (therefore no tests in the group run), yet the `setUpAll` and `tearDownAll` callbacks are executed.  I admit that having no tests is a degenerate case and that I am being pedantic.  But also consider if we had:

```dart
    setUpAll(() {
      print('setUpAll');
      throw Exception('fail');
    });
```
so now the `setUpAll` callback fails, and even if there were tests, they'd never run, but the `tearDownAll` callback still executes anyway.  It seems wrong to me to execute the `tearDownAll` callback if `setUpAll` fails (similar to how you don't call destructors on objects if their constructors failed).

I ran into this during a null-safety migration where some migrated test code did:

```dart
late Directory tempDir;
setUpAll(() {
  if (verifyProperlyConfiguredEnvironment()) {
    throw Exception('You need to configure your environment, yada yada yada.');
  }
  tempDir = Directory.systemTemp.createTempSync('foo');
});

tearDownAll(() {
  tempDir.deleteSync(recursive: true);
});
```
which on failure complains about `tempDir` being uninitialized.

**Comment by jakemac53** [comment]
What about this case though?

```
late Directory tempDir;
setUpAll(() {
  tempDir = Directory.systemTemp.createTempSync('foo');
  if (verifyProperlyConfiguredEnvironment()) {
    throw Exception('You need to configure your environment, yada yada yada.');
  }
});

tearDownAll(() {
  tempDir.deleteSync(recursive: true);
});
```

Now it is important that `tearDownAll` does in fact run in order to clean up artifacts. In practice you may also be cancelling stream listeners, sub-processes, etc which are more important than cleaning up a temp dir.

The issue you describe is quite specific to `late` variables. Any access to a `late` variable has the possibility of failing at runtime.

**Comment by jamesderlin** [comment]
> What about this case though?

If `setUpAll`'s callback fails and does work that needs to be cleaned up, the contract should be that the callback is responsible for cleaning up after itself.  If `setUpAll`'s callback fails, the callback is in the best position to determine exactly what should be cleaned up and how, especially if it acquires multiple resources with multiple points for failure and should do cleanup in some particular order.  That's no different than a C++ constructor that does work before failing.

**Comment by jakemac53** [comment]
In general though cleaning up is the job of `tearDown`/`tearDownAll`. That is the contract in the context of the test framework. Putting the cleanup code in both places doesn't seem right, and most people just aren't going to do that.

Again I think the real issue here is the usage of `late` in this context. By using that feature, you are saying "I promise I will initialize this before I use it". But in this context you didn't initialize it (due to an exception preventing that code from being reached).

I suggest not using `late` unless you are 100% sure that you will in fact initialize the variable.

**Comment by jakemac53** [comment]
> At a minimum, the documentation to `tearDownAll` should be updated to indicate that the callback will be run even if `setUpAll` fails. The current claim that it won't be executed if no tests run is not true.

I agree the docs should be updated. The `setUpAll` and `tearDownAll` methods always run, even with no tests present.
